src/auth.js

```diff
@@ -117,20 +132,147 @@ class Auth {
     return true;
   }
 
+  /**
+   * Creates a refresh token for a user and stores it on the server.
+   *
+   * @private
+   * @param {Request} req
+   * @param {string} username
+   * @param {array} groups
+   * @returns {string}
+   */
+  createRefreshToken(req, username, groups) {
+    const refreshToken = jwt.sign(
+      {username, groups: JSON.stringify(groups)},
+      this.refreshTokenSecret
+    );
+
+    req.session.refreshTokens.push(refreshToken);
```

Comment:
I don't quite understand why these go into the session. Could you elaborate ?

Reply:
The refresh tokens go into the session to ensure their validity. When a user logs out, the refresh token is deleted from the session. Thus if a bad actor gets a hold of a user's refresh token, the token becomes useless once the user logs out. This is an additional safeguard due to the fact that refresh tokens do not expire automatically, and we do not want an impostor having permanent access to a user's account.

Let me know if that makes sense, I can try to find a better way to explain it if not. I am also open to alternative ideas if you have something else in mind.

Reply:
Hm. But this only works if the user actually logs out. Tokens should be nuked automatically.

Which is where the storage talked about in the previous review comes into play. These tokens should not live here imo.

Reply:
Is `req.session` not the storage you discussed? I transitioned to this method because I believe `req.session` is tied to the `loki-connect` storage as we discussed before, with the ability to be stored on the disk for persistence and the ability to use Redis as a drop-in replacement as you mentioned.

If I am misunderstanding and I am supposed to access the `express-session` storage in a different manner, please correct me.

But yes, this does only work if the user logs out. I do not have the refresh tokens automatically expiring because there is no good way to keep the user logged in if that is the case, requests would just start to fail in the middle of their session if the refresh token expired because they would be unable to obtain a new access token. I am open to alternative ways to work around this, I was just not able to think of any good ones.

Reply:
> Is req.session not the storage you discussed? I transitioned to this method because I believe req.session is tied to the loki-connect storage as we discussed before, with the ability to be stored on the disk for persistence and the ability to use Redis as a drop-in replacement as you mentioned.

Ah, but now you're hooking into the cookies, which is what we're moving away from :)

My idea was that you'd create an abstraction for JWT token storage where you'd have an interface that could use something like lokijs (which is already a dependency we have from cooke storage) as a default storage.

So something like:

```js
class LokiTokenStorage { // This is basically the interface needed :
  constructor(core, options) {}
  async init() {}
  async find() {}
  async create() {}
  async remove() {}
}

class Auth {
  constructor(core) {
    // This could be customized via the service provider or whatver, but the gist is:
    this.storage = new LokiTokenStorage(core, /* you can pass options here */)
  }
  async init() {
    await this.storage.init()
  }
}
```

Hope that makes things a bit more clear.

Reply:
> I am open to alternative ways to work around this, I was just not able to think of any good ones.

I have a few ideas, but let's pick that up once we have figured out this store thing.

Reply:
> Ah, but now you're hooking into the cookies, which is what we're moving away from :)

Can you elaborate? From my understanding, data stored in `req.session` is stored in the `LokiStore` created by `connect-loki`, which is persistent. I do not see the connection to cookies.

> My idea was that you'd create an abstraction for JWT token storage where you'd have an interface that could use something like lokijs (which is already a dependency we have from cooke storage) as a default storage.

I do not understand the need for/purpose of the additional recommended `LokiTokenStorage` class. The data we are interested in is stored in `./session-store.db` where it persists. Could you elaborate on what you would want those functions to accomplish?

Reply:
The current integration is with `express-session`. This puts a unique ID ("session id") into a cookie.

So on every request, the session id is looked up in loki, and the data is brought back and put into `req.session`.

When you mutate the `req.session`, it gets stored back into loki based on that ID.

---

JWT however is immutable. You put the session data into the token on login, and that's it. They don't really need to be stored anywhere because they're signed and has a very small lifetime.

But with this refresh token mechanism there needs to be some kind of long-term storage available so that a server shutdown does not introduce the need to do a new login.

Reply:
> But with this refresh token mechanism there needs to be some kind of long-term storage available

And also to be able to use the refresh tokens between different instances of the server if it's scaled.

Reply:
In order to make this transition, the entire `express-session` needs to to go, and the token data needs to go into `req.session` just as it is today so that there's no breakage in the APIs.

This also means that you can't use `req.session` to store arbitrary data any more. There's need for a new abstraction

> Could you elaborate on what you would want those functions to accomplish?

The example class interface I wrote would be like a repository to look up and manage tokens (in this case the refresh ones).

Reply:
> the entire express-session needs to to go

Does that mean you want `createSession` to go as well? I'm not sure what purpose the function would serve without `express-session`, but I want to make sure before I get rid of it entirely.

> (which means no more cookies)

Does this mean you want to get rid of the client-side cookie that stores the refresh token as well? I don't see how removing `express-session` has any impact on that functionality, `res.status(200).json(profile)` will continue to return the same data that can be stored in that cookie to ensure continued availability of access tokens to the user without reauthentication.

> the token data needs to go into req.session just as it is today so that there's no breakage in the APIs. This also means that you can't use req.session to store arbitrary data any more.

I'm confused on this part, can you explain what you mean? I apologize for my ignorance, I haven't implemented a JWT auth mechanism from scratch before so this is a learning process for me.

Reply:
> Does that mean you want createSession to go as well? I'm not sure what purpose the function would serve without express-session, but I want to make sure before I get rid of it entirely.

Anything related to this old session stuff has to go.

> Does this mean you want to get rid of the client-side cookie that stores the refresh token as well?

Tokens should not really be stored on the client in any way. These can just live in memory as a variable somewhere.

> has any impact on that functionality

It actually has. Look at my notes below. Manipulating `req.session` is a valid thing to do when there are cookies available.

> I'm confused on this part, can you explain what you mean? I apologize for my ignorance, I haven't implemented a JWT auth mechanism from scratch before so this is a learning process for me.

When you log into a system like this, you get a token. This token now contains what the server was doing previously, i.e.:

```
{
  "username": "foo",
  "id": 123,
  "groups": ["a", "b", "c"]
}
```

The client has to send this token on every request in order to be authorized, and the `req.session` on the server has to be populated with this data in some way (it's under `req.session.user` today).

Reply:
> How do you plan on tying the tokens to a particular user then, if not by storing the user's refresh token on the client side so it is tied to that particular machine? Also related, isn't this functionality necessary to implement https://github.com/os-js/osjs-client/issues/88?

Well, to enable that kind of functionality it would have to go into the localStorage or something. Refresh tokens come with some downsides, but there's no way around that :/

> Are you saying you just want the data that was stored in req.session.user to be stored in req.session now instead?

`req.session.user` as it is today. The `req.session` is just a "namespace" or whatever you'd call it where the `express-session` places stuff on every request.

Reply:
I don't know how to continue storing persistent data in `req.session` without `express-session`, do you have any thoughts? Everything else is just about entirely implemented.

Reply:
Also, I checked through the manual again, and it doesn't seem like there's any examples on how to do this, but it says that  `req.session` is accessible. Pretty vague wording there, hehe.

Reply:
I think there's some confusion here.

Let me start over....

This will implement JWT mechanism in the authentication module and swaps out the authorization for the API with a new check mechanic.

This means anything related to cookies must go entirely. And this JWT needs to go into `req.session.user`.

Now... since the cookies are gone, it's no longer possible to use `req.session` as an arbitrary storage.

This is a semi-breaking change. Because I don't know how many people do use `req.session` outside just reading the user data (aka `req.session.user`).

Reply:
> an alternative method of retrieving the data of interest via the JWTs, if that would work for you.

But in this case I feel like it's better just to stick with cookies because it already does *all the things* :man_shrugging:

Reply:
> Because I don't know how many people do use req.session outside just reading the user data 

I might just ask some of the guys I know that runs this in production. One of them also has built a JWT mechanism via a login portal and custom adapters, and my guess is that they just put everything they need into the token upon login and force re-logins if the information that goes into it changes somehow.

Reply:
... continuation from last comment now that I've made some discoveries. Seems like `req.session` is potentially being used quite a bit.

That's a shame. I was hoping this could be achieved in a total stateless way and not to hit a database. But considering that there might be secrets stored in the session etc. it can't just go into a token.

So there is a need for some kind of storage here after all. And it has to be compatible with the old session API stuff.

Reply:
> But then I feel like it's better just to stick with cookies because it already does *all the things*

That's what I said originally, the whole point of the JWTs was to transition the cookie from cleartext username and password combination to a refresh token, which would minimize the security risk if leaked because a refresh token can be invalidated much more easily than a username and password can.

I didn't see any reason for getting rid of `express-session`, as it can easily be used in conjunction with my JWT method. I only got rid of it because of this:

> In order to make this transition, the entire `express-session` needs to to go

The inclusion of JWTs and the usage of `express-session` are not mutually exclusive

Reply:
> I didn't see any reason for getting rid of express-session, as it can easily be used in conjunction with my JWT method.

But now you have *both* a session cookie and a JWT. That kind of defeats the purpose of transitioning to this!